code/modules/cargo/randomstock.dm

```diff
@@ -45,15 +45,42 @@ STOCK_ITEM_COMMON(bees, 2)
 
 */
 
-// This datum doesn't actually do anything beyond hold the setup_cargo_stock proc. Defining it as a global proc causes things to break.
-/var/datum/cargo_master/cargo_master = new
-
-
-// Called in misc_early to generate the cargo lists (as Atoms runs before Cargo).
-/datum/cargo_master/proc/setup_cargo_stock()
-
-
-// These lists are populated by the files in `./random_stock`.
+// The cargo_master datum is used to hold metadata about the cargo spawners. All subtypes are instantiated in misc_early to generate the cargo lists (as Atoms runs before Cargo).
+// This also means that if you want to do something fancy with the spawning probability you can directly create a custom sub-datum of this
+// instead of using the macros and override the get_probability proc or even the register_spawner proc.
+/datum/cargo_master
+	var/category = null
+	var/probability = 0
+	var/spawner_proc = null
+
+/datum/cargo_master/proc/get_probability()
+	return src.probability
+
+/datum/cargo_master/proc/register_spawner()
+	var/P = src.get_probability()
+	if(P <= 0)
+		return 0
+	switch(src.category)
+		if("common")
+			return global.random_stock_common[src.spawner_proc] = P
+		if("uncommon")
+			return global.random_stock_uncommon[src.spawner_proc] = P
+		if("rare")
+			return global.random_stock_rare[src.spawner_proc] = P
+		if("large")
+			return global.random_stock_large[src.spawner_proc] = P
+	throw EXCEPTION("Cargo spawner definition '[src.type]' has invalid category '[src.category]'. Please fix your definition.")
+
+/proc/setup_cargo_spawn_lists()
+	var/i
+	for(var/type in subtypesof(/datum/cargo_master))
+		var/datum/cargo_master/def = new type()
+		def.register_spawner()
+		qdel(def)
```

Comment:
Doesn't `def` only exist in this scope, so it'll be cleaned up automagically, so we don't need this qdel?

Reply:
There are no external references held to it so Byond should probably clean it eventually. I thought I'd help. Is there a chance that this would break something, or is it just unnecessary?

Reply:
It's just unnecessary.